File: backend/app/routers/stocks.py
from typing import Literal, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.core.auth import get_current_user
from app.core.database import get_db
from app.models.user_platform import AppUser
from app.schemas.stock import (
    MarketType,
    RecommendationType,
    StockEnrichResponse,
    StockEnrichStatusResponse,
    StockKLineResponse,
    StockAnalysis,
    StockDetail,
    StockQARequest,
    StockQAResponse,
    StockListResponse,
    SectorRotationResponse,
    StockSnapshot,
    StockSyncResponse,
    StockTradeReviewCreate,
    StockTradeReviewItem,
    StockTradeReviewResponse,
    StockTradeReviewUpdate,
)
from app.services.enrichment_service import enrich_stock_universe, get_enrichment_status
from app.services.stock_service import (
    ask_stock_question,
    get_sector_rotation_summary,
    get_stock_analysis,
    get_stock_kline,
    get_stock_detail,
    get_stock_list,
    get_stock_snapshot,
    sync_stock_universe,
)
from app.services.trade_review_service import create_stock_trade_review, get_stock_trade_reviews, update_stock_trade_review

router = APIRouter(prefix="/stocks", tags=["stocks"])


@router.get("", response_model=StockListResponse)
def list_stocks(
    analyzed: Optional[bool] = Query(default=None),
    market: Optional[MarketType] = Query(default=None),
    board: Optional[str] = Query(default=None),
    exchange: Optional[str] = Query(default=None),
    industry: Optional[str] = Query(default=None),
    concept: Optional[str] = Query(default=None),
    tag: Optional[str] = Query(default=None),
    recommendation: Optional[RecommendationType] = Query(default=None),
    dividend_only: Optional[bool] = Query(default=None),
    dividend_years_min: Optional[int] = Query(default=None, ge=1, le=20),
    dividend_yield_min: Optional[float] = Query(default=None, ge=0, le=50),
    ex_dividend_soon: Optional[bool] = Query(default=None),
    market_cap_min: Optional[float] = Query(default=None, ge=0, le=1000000),
    market_cap_max: Optional[float] = Query(default=None, ge=0, le=1000000),
    price_min: Optional[float] = Query(default=None, ge=0, le=1000000),
    price_max: Optional[float] = Query(default=None, ge=0, le=1000000),
    pe_max: Optional[float] = Query(default=None, ge=0, le=500),
    net_profit_min: Optional[float] = Query(default=None, ge=0, le=1000000),
    revenue_min: Optional[float] = Query(default=None, ge=0, le=1000000),
    revenue_growth_min: Optional[float] = Query(default=None, ge=-100, le=300),
    revenue_growth_qoq_min: Optional[float] = Query(default=None, ge=-100, le=300),
    profit_growth_min: Optional[float] = Query(default=None, ge=-100, le=300),
    profit_growth_qoq_min: Optional[float] = Query(default=None, ge=-100, le=300),
    gross_margin_min: Optional[float] = Query(default=None, ge=0, le=100),
    net_margin_min: Optional[float] = Query(default=None, ge=-100, le=100),
    roe_min: Optional[float] = Query(default=None, ge=-100, le=100),
    debt_ratio_max: Optional[float] = Query(default=None, ge=0, le=100),
    exclude_st: Optional[bool] = Query(default=None),
    score_min: Optional[int] = Query(default=None, ge=1, le=99),
    score_max: Optional[int] = Query(default=None, ge=1, le=99),
    change_pct_min: Optional[float] = Query(default=None, ge=-30, le=30),
    change_pct_max: Optional[float] = Query(default=None, ge=-30, le=30),
    prev_limit_up: Optional[bool] = Query(default=None),
    prev_limit_down: Optional[bool] = Query(default=None),
    q: Optional[str] = Query(default=None),
    sort_by: Literal["score", "change_pct", "price", "dividend_years", "dividend_yield"] = Query(default="score"),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=50, ge=10, le=200),
    include_meta: bool = Query(default=True),
    db: Session = Depends(get_db),
) -> StockListResponse:
    return get_stock_list(
        db=db,
        analyzed=analyzed,
        market=market,
        board=board,
        exchange=exchange,
        industry=industry,
        concept=concept,
        tag=tag,
        recommendation=recommendation,
        dividend_only=dividend_only,
        dividend_years_min=dividend_years_min,
        dividend_yield_min=dividend_yield_min,
        ex_dividend_soon=ex_dividend_soon,
        market_cap_min=market_cap_min,
        market_cap_max=market_cap_max,
        price_min=price_min,
        price_max=price_max,
        pe_max=pe_max,
        net_profit_min=net_profit_min,
        revenue_min=revenue_min,
        revenue_growth_min=revenue_growth_min,
        revenue_growth_qoq_min=revenue_growth_qoq_min,
        profit_growth_min=profit_growth_min,
        profit_growth_qoq_min=profit_growth_qoq_min,
        gross_margin_min=gross_margin_min,
        net_margin_min=net_margin_min,
        roe_min=roe_min,
        debt_ratio_max=debt_ratio_max,
        exclude_st=exclude_st,
        score_min=score_min,
        score_max=score_max,
        change_pct_min=change_pct_min,
        change_pct_max=change_pct_max,
        prev_limit_up=prev_limit_up,
        prev_limit_down=prev_limit_down,
        q=q,
        sort_by=sort_by,
        page=page,
        page_size=page_size,
        include_meta=include_meta,
    )


@router.get("/sectors/rotation", response_model=SectorRotationResponse)
def get_sector_rotation_endpoint(
    market: Optional[MarketType] = Query(default=None),
    top_n: int = Query(default=8, ge=3, le=15),
    db: Session = Depends(get_db),
) -> SectorRotationResponse:
    return get_sector_rotation_summary(db=db, market=market, top_n=top_n)


@router.post("/sync", response_model=StockSyncResponse)
def sync_stocks(
    force: bool = Query(default=False),
    db: Session = Depends(get_db),
) -> StockSyncResponse:
    return sync_stock_universe(db=db, force=force)


@router.post("/enrich", response_model=StockEnrichResponse)
def enrich_stocks(
    force: bool = Query(default=False),
    market: Optional[MarketType] = Query(default=None),
    limit: Optional[int] = Query(default=None, ge=1, le=10000),
    sleep_ms: int = Query(default=120, ge=0, le=3000),
    db: Session = Depends(get_db),
) -> StockEnrichResponse:
    return enrich_stock_universe(
        db=db,
        force=force,
        market=market,
        limit=limit,
        sleep_ms=sleep_ms,
    )


@router.get("/enrich/status", response_model=StockEnrichStatusResponse)
def get_enrich_status(db: Session = Depends(get_db)) -> StockEnrichStatusResponse:
    return get_enrichment_status(db=db)


@router.get("/{symbol}/reviews", response_model=StockTradeReviewResponse)
def get_stock_trade_reviews_endpoint(
    symbol: str,
    user: AppUser = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> StockTradeReviewResponse:
    response = get_stock_trade_reviews(db=db, symbol=symbol, user_id=user.id, include_all=(user.role == "admin"))
    if response is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="股票不存在")
    return response


@router.post("/{symbol}/reviews", response_model=StockTradeReviewItem, status_code=status.HTTP_201_CREATED)
def create_stock_trade_review_endpoint(
    symbol: str,
    payload: StockTradeReviewCreate,
    user: AppUser = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> StockTradeReviewItem:
    try:
        response = create_stock_trade_review(db=db, symbol=symbol, payload=payload, user_id=user.id)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    if response is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="股票不存在")
    return response


@router.patch("/{symbol}/reviews/{review_id}", response_model=StockTradeReviewItem)
def update_stock_trade_review_endpoint(
    symbol: str,
    review_id: int,
    payload: StockTradeReviewUpdate,
    user: AppUser = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> StockTradeReviewItem:
    try:
        response = update_stock_trade_review(
            db=db,
            symbol=symbol,
            review_id=review_id,
            payload=payload,
            user_id=user.id,
            can_manage_all=(user.role == "admin"),
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    if response is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="股票或复盘记录不存在")
    return response


@router.get("/{symbol}/kline", response_model=StockKLineResponse)
def get_stock_kline_endpoint(
    symbol: str,
    period: Literal["1mo", "3mo", "6mo", "1y", "5y"] = Query(default="6mo"),
    interval: Literal["1d", "1h"] = Query(default="1d"),
    db: Session = Depends(get_db),
) -> StockKLineResponse:
    response = get_stock_kline(db=db, symbol=symbol, period=period, interval=interval)
    if response is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="股票不存在")
    return response


@router.post("/{symbol}/qa", response_model=StockQAResponse)
def ask_stock_question_endpoint(
    symbol: str,
    payload: StockQARequest,
    db: Session = Depends(get_db),
) -> StockQAResponse:
    response = ask_stock_question(db=db, symbol=symbol, question=payload.question, history=[item.model_dump() for item in payload.history])
    if response is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="股票不存在")
    return response


@router.get("/{symbol}", response_model=StockDetail)
def get_stock_detail_endpoint(symbol: str, db: Session = Depends(get_db)) -> StockDetail:
    stock = get_stock_detail(db=db, symbol=symbol)
    if stock is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="股票不存在")
    return stock


@router.get("/{symbol}/analysis", response_model=StockAnalysis)
def get_stock_analysis_endpoint(symbol: str, db: Session = Depends(get_db)) -> StockAnalysis:
    analysis = get_stock_analysis(db=db, symbol=symbol)
    if analysis is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="股票不存在")
    return analysis


@router.get("/{symbol}/snapshot", response_model=StockSnapshot)
def get_stock_snapshot_endpoint(symbol: str, db: Session = Depends(get_db)) -> StockSnapshot:
    snapshot = get_stock_snapshot(db=db, symbol=symbol)
    if snapshot is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="股票不存在")
    return snapshot

File: frontend/src/api/stocks.ts
import { request } from "./client";
import type {
  DashboardSummary,
  MarketType,
  RecommendationType,
  ReportDetail,
  ReportListItem,
  SectorRotationResponse,
  StockAnalysis,
  StockDetail,
  StockSnapshot,
  StockQARequest,
  StockQAResponse,
  StockEnrichResponse,
  StockEnrichStatusResponse,
  StockKLineResponse,
  StockListResponse,
  StockTradeReviewCreateRequest,
  StockTradeReviewItem,
  StockTradeReviewResponse,
  StockSyncResponse,
  StockTradeReviewUpdateRequest,
} from "../types/stock";

interface StockListQuery {
  analyzed?: boolean;
  market?: MarketType;
  board?: string;
  exchange?: string;
  industry?: string;
  concept?: string;
  tag?: string;
  recommendation?: RecommendationType;
  dividend_only?: boolean;
  dividend_years_min?: number;
  dividend_yield_min?: number;
  ex_dividend_soon?: boolean;
  market_cap_min?: number;
  market_cap_max?: number;
  price_min?: number;
  price_max?: number;
  pe_max?: number;
  net_profit_min?: number;
  revenue_min?: number;
  revenue_growth_min?: number;
  revenue_growth_qoq_min?: number;
  profit_growth_min?: number;
  profit_growth_qoq_min?: number;
  gross_margin_min?: number;
  net_margin_min?: number;
  roe_min?: number;
  debt_ratio_max?: number;
  exclude_st?: boolean;
  score_min?: number;
  score_max?: number;
  change_pct_min?: number;
  change_pct_max?: number;
  prev_limit_up?: boolean;
  prev_limit_down?: boolean;
  q?: string;
  sort_by?: "score" | "change_pct" | "price" | "dividend_years" | "dividend_yield";
  page?: number;
  page_size?: number;
  include_meta?: boolean;
}

interface RequestOptions {
  signal?: AbortSignal;
}

interface StockEnrichQuery {
  force?: boolean;
  market?: MarketType;
  limit?: number;
  sleep_ms?: number;
}

export async function listStocks(query: StockListQuery = {}, options: RequestOptions = {}): Promise<StockListResponse> {
  const params = new URLSearchParams();

  if (query.analyzed !== undefined) {
    params.set("analyzed", String(query.analyzed));
  }
  if (query.market) {
    params.set("market", query.market);
  }
  if (query.board) {
    params.set("board", query.board);
  }
  if (query.exchange) {
    params.set("exchange", query.exchange);
  }
  if (query.industry) {
    params.set("industry", query.industry);
  }
  if (query.concept) {
    params.set("concept", query.concept);
  }
  if (query.tag) {
    params.set("tag", query.tag);
  }
  if (query.recommendation) {
    params.set("recommendation", query.recommendation);
  }
  if (query.dividend_only) {
    params.set("dividend_only", "true");
  }
  if (query.dividend_years_min !== undefined) {
    params.set("dividend_years_min", String(query.dividend_years_min));
  }
  if (query.dividend_yield_min !== undefined) {
    params.set("dividend_yield_min", String(query.dividend_yield_min));
  }
  if (query.ex_dividend_soon) {
    params.set("ex_dividend_soon", "true");
  }
  if (query.market_cap_min !== undefined) {
    params.set("market_cap_min", String(query.market_cap_min));
  }
  if (query.market_cap_max !== undefined) {
    params.set("market_cap_max", String(query.market_cap_max));
  }
  if (query.price_min !== undefined) {
    params.set("price_min", String(query.price_min));
  }
  if (query.price_max !== undefined) {
    params.set("price_max", String(query.price_max));
  }
  if (query.pe_max !== undefined) {
    params.set("pe_max", String(query.pe_max));
  }
  if (query.net_profit_min !== undefined) {
    params.set("net_profit_min", String(query.net_profit_min));
  }
  if (query.revenue_min !== undefined) {
    params.set("revenue_min", String(query.revenue_min));
  }
  if (query.revenue_growth_min !== undefined) {
    params.set("revenue_growth_min", String(query.revenue_growth_min));
  }
  if (query.revenue_growth_qoq_min !== undefined) {
    params.set("revenue_growth_qoq_min", String(query.revenue_growth_qoq_min));
  }
  if (query.profit_growth_min !== undefined) {
    params.set("profit_growth_min", String(query.profit_growth_min));
  }
  if (query.profit_growth_qoq_min !== undefined) {
    params.set("profit_growth_qoq_min", String(query.profit_growth_qoq_min));
  }
  if (query.gross_margin_min !== undefined) {
    params.set("gross_margin_min", String(query.gross_margin_min));
  }
  if (query.net_margin_min !== undefined) {
    params.set("net_margin_min", String(query.net_margin_min));
  }
  if (query.roe_min !== undefined) {
    params.set("roe_min", String(query.roe_min));
  }
  if (query.debt_ratio_max !== undefined) {
    params.set("debt_ratio_max", String(query.debt_ratio_max));
  }
  if (query.exclude_st) {
    params.set("exclude_st", "true");
  }
  if (query.score_min !== undefined) {
    params.set("score_min", String(query.score_min));
  }
  if (query.score_max !== undefined) {
    params.set("score_max", String(query.score_max));
  }
  if (query.change_pct_min !== undefined) {
    params.set("change_pct_min", String(query.change_pct_min));
  }
  if (query.change_pct_max !== undefined) {
    params.set("change_pct_max", String(query.change_pct_max));
  }
  if (query.prev_limit_up) {
    params.set("prev_limit_up", "true");
  }
  if (query.prev_limit_down) {
    params.set("prev_limit_down", "true");
  }
  if (query.q) {
    params.set("q", query.q);
  }
  if (query.sort_by) {
    params.set("sort_by", query.sort_by);
  }
  if (query.page) {
    params.set("page", String(query.page));
  }
  if (query.page_size) {
    params.set("page_size", String(query.page_size));
  }
  if (query.include_meta === false) {
    params.set("include_meta", "false");
  }

  return request<StockListResponse>({
    path: `/stocks${params.toString() ? `?${params.toString()}` : ""}`,
    method: "GET",
    signal: options.signal,
  });
}

export async function syncStockUniverse(force = false): Promise<StockSyncResponse> {
  const params = new URLSearchParams();
  if (force) {
    params.set("force", "true");
  }

  return request<StockSyncResponse>({
    path: `/stocks/sync${params.toString() ? `?${params.toString()}` : ""}`,
    method: "POST",
  });
}

export async function enrichStockUniverse(query: StockEnrichQuery = {}): Promise<StockEnrichResponse> {
  const params = new URLSearchParams();

  if (query.force) {
    params.set("force", "true");
  }
  if (query.market) {
    params.set("market", query.market);
  }
  if (query.limit !== undefined) {
    params.set("limit", String(query.limit));
  }
  if (query.sleep_ms !== undefined) {
    params.set("sleep_ms", String(query.sleep_ms));
  }

  return request<StockEnrichResponse>({
    path: `/stocks/enrich${params.toString() ? `?${params.toString()}` : ""}`,
    method: "POST",
  });
}

export async function getStockEnrichStatus(): Promise<StockEnrichStatusResponse> {
  return request<StockEnrichStatusResponse>({
    path: "/stocks/enrich/status",
    method: "GET",
  });
}

export async function getStockDetail(symbol: string): Promise<StockDetail> {
  return request<StockDetail>({
    path: `/stocks/${encodeURIComponent(symbol)}`,
    method: "GET",
  });
}

export async function getStockAnalysis(symbol: string): Promise<StockAnalysis> {
  return request<StockAnalysis>({
    path: `/stocks/${encodeURIComponent(symbol)}/analysis`,
    method: "GET",
  });
}

export async function getStockSnapshot(symbol: string): Promise<StockSnapshot> {
  return request<StockSnapshot>({
    path: `/stocks/${encodeURIComponent(symbol)}/snapshot`,
    method: "GET",
  });
}

export async function askStockQuestion(symbol: string, payload: StockQARequest, options: RequestOptions = {}): Promise<StockQAResponse> {
  return request<StockQAResponse>({
    path: `/stocks/${encodeURIComponent(symbol)}/qa`,
    method: "POST",
    body: JSON.stringify(payload),
    signal: options.signal,
  });
}

export async function getStockKLine(
  symbol: string,
  period: "1mo" | "3mo" | "6mo" | "1y" | "5y" = "6mo",
  interval: "1d" | "1h" = "1d"
): Promise<StockKLineResponse> {
  const params = new URLSearchParams();
  params.set("period", period);
  params.set("interval", interval);

  return request<StockKLineResponse>({
    path: `/stocks/${encodeURIComponent(symbol)}/kline?${params.toString()}`,
    method: "GET",
  });
}

export async function getStockTradeReviews(symbol: string): Promise<StockTradeReviewResponse> {
  return request<StockTradeReviewResponse>({
    path: `/stocks/${encodeURIComponent(symbol)}/reviews`,
    method: "GET",
  });
}

export async function createStockTradeReview(
  symbol: string,
  payload: StockTradeReviewCreateRequest
): Promise<StockTradeReviewItem> {
  return request<StockTradeReviewItem>({
    path: `/stocks/${encodeURIComponent(symbol)}/reviews`,
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function updateStockTradeReview(
  symbol: string,
  reviewId: number,
  payload: StockTradeReviewUpdateRequest
): Promise<StockTradeReviewItem> {
  return request<StockTradeReviewItem>({
    path: `/stocks/${encodeURIComponent(symbol)}/reviews/${reviewId}`,
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export async function getDashboardSummary(): Promise<DashboardSummary> {
  return request<DashboardSummary>({
    path: "/dashboard/summary",
    method: "GET",
  });
}

export async function getSectorRotation(market?: MarketType, topN = 8, options: RequestOptions = {}): Promise<SectorRotationResponse> {
  const params = new URLSearchParams();
  if (market) {
    params.set("market", market);
  }
  params.set("top_n", String(topN));

  return request<SectorRotationResponse>({
    path: `/stocks/sectors/rotation${params.toString() ? `?${params.toString()}` : ""}`,
    method: "GET",
    signal: options.signal,
  });
}

export async function listReports(q?: string): Promise<ReportListItem[]> {
  const params = new URLSearchParams();

  if (q?.trim()) {
    params.set("q", q.trim());
  }

  return request<ReportListItem[]>({
    path: `/reports${params.toString() ? `?${params.toString()}` : ""}`,
    method: "GET",
  });
}

export async function getReportDetail(symbol: string): Promise<ReportDetail> {
  return request<ReportDetail>({
    path: `/reports/${encodeURIComponent(symbol)}`,
    method: "GET",
  });
}
